### ISSUE TITLE
Rule: Shadowed names

### ISSUE THREAD

**Issue by Mauin** [detekt/detekt#240]
The Kotlin compiler already prints a warning for each shadowed name. Should Detekt also detect these and report them?

Example:
```
override fun bindItemView(position: Int, view: View, list: List<Item>) {        
    setOnClickListener { view -> handleClick(view) }
}
```
The view in the `setOnClickListener` lambda is shadowing the fun parameter.

**Comment by 3flex** [comment]
Not related to this issue, just want to ask - in general, should detekt replicate checks that are already built in to the compiler? I'd suggest not to reduce duplication of effort. #269 is another that is checked by the compiler already.

**Comment by arturbosch** [comment]
On the one hand no we shouldn't implement them as the kotlin compiler does a better job at this but on the other hand the compiler does not break the build and detekts result can be piped to slack etc

**Comment by 3flex** [comment]
Since Kotlin 1.2 can use `-Werror` to fail the build on compiler warnings, or with Gradle
```groovy
compileKotlin {
    kotlinOptions.allWarningsAsErrors = true
}
```

I can't speak to piping the results elsewhere though and you don't have the options to configure rules individually like in detekt.

**Comment by vanniktech** [comment]
Oh that's interesting. If this really works I'd let the compiler do the job and close all related issues that can be solved with that.

**Comment by vanniktech** [comment]
For Android you'll need to do this:

```groovy
tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
  kotlinOptions {
      allWarningsAsErrors = true
  }
}
```

However it's not possible to mark this:

```
name.setOnClickListener { name ->
  name.alpha = 0.5f
}
```

**Comment by Mauin** [comment]
That's an interesting thought. But is it possible to tune the compiler to only report certain types of warnings as errors? Also in a CI environment I might want to split a build step and a static analysis step and have both report separately to make it easier to understand what failed in my build.
I still see some value of having this rule in detekt.

**Comment by pcl** [comment]
After having played around with `-Werror` in a non-trivial codebase, I've come to the conclusion that a `NoNameShadowing` rule is valuable for inclusion in detekt, for two reasons:

1. `-Werror` is too blunt of an instrument. In particular, this results in compilation failures on deprecated API utilization. Sometimes, deprecated APIs are unavoidable for some period of time, so those often need to remain as warnings.

2. Kotlin's name shadowing warnings are too lenient. For example, a class-scoped `val` can be shadowed without warning. `kotlinc -Werror` produces no warnings for this input:

```kotlin
class Foo {
    val foo = "aoeu"
    fun yo() {
        val foo = "snth"
        System.out.println("foo: ${foo}")
    }
}
```

**Comment by schalkms** [comment]
@pcl thanks for your efforts!
Please submit a PR. I will review it then.

**Comment by pcl** [comment]
Brief update: I picked up the PR again late last week, and have it pretty much done. I've run it through a number of the suggested sample projects, and came across a few use cases that it's alerting on that I think are legit patterns. I hope to have a PR ready for you to take a look at later this week.

Meanwhile, a question: I'd like to be able to raise an alert on nested implicit `it`s, like so:

```kotlin
listOf(listOf(0, 1), listOf(2)).forEach {
    it.forEach { // 'it' #1
        System.out.println(it) // 'it' #2
    }
}
```

However, the baseline visitor pattern that the `detekt` rules use do not have any knowledge of the implicit `it` in those cases. Essentially, to detect that sort of issue, I need to plug into a compilation phase that is aware of the length of the parameter list of the `forEach` closure. Is there such a phase available within the existing infrastructure?

**Comment by arturbosch** [comment]
@pcl yes please submit a PR!
You can't build detekt with gradle 4.9 because we are using the kotlin-dsl which got breaking changes in 4.7. We are still migrating [here](https://github.com/arturbosch/detekt/pull/1017). So please use gradle 4.6 to build and test your changes. (If you are using sdkman switching gradle versions is easy :))

**Comment by arturbosch** [comment]
Heyho, for your example I just tried this with the debugger:
```
(((((ktFile.children[2] as KtNamedFunction).bodyExpression.children[0] as KtDotQualifiedExpression).selectorExpression as KtCallExpression).lambdaArguments[0] as KtLambdaArgument).getLambdaExpression() as KtLambdaExpression).valueParameters
```

For `forEach { // just body }` I get an empty valueParameters list and for `forEach { it -> // body }` you will get the it as an ValueParameter. In the LambdaBody you can check if there is a reference to `it` somewhere. Hope that helps.

**Comment by pcl** [comment]
Yeah, that works to get the parameter list as declared in the source code. But how do you figure out if the lambda would have a single param or not? In other words, how do you determine that the called function accepts a one-arg lambda vs. a zero-arg lambda? Here's an example using a `forEach` block for the implicit `it` and a `Runnable.run()`, which has no implicit arg and thus should be legal:

```kotlin
val list = listOf("foo", "bar")
val executor = ...
list.forEach {
    val thread = Thread { // create a closure for Thread's Runnable-accepting constructor
        System.err.println("${it}") // this 'it' is implicitly defined for 'forEach'
    }
}
```

So, I think I'd need to be able to do at least some basic class loading to look at `Thread`'s constructor in the above example.
